File: src/bot.py

### Discord Bot for Emergency Hamburg Private Server Information
### Coded by Unl1veabl3

# ---------- Imports ----------

import discord
from discord import app_commands
from discord.ui import View, Button
import requests
import time

# ---------- Config ----------
TOKEN = "YOUR_DISCORD_BOT_TOKEN_HERE"

# Set this to an integer if you want guild-only commands
# Set to None for global slash commands
GUILD_ID = None
# Example:
# GUILD_ID = 1450502148100980796


# ---------- API ----------
API_URL = "https://api.emergency-hamburg.com/public/servers"


# ---------- helpers ----------

def find_server_by_owner_id(owner_id: int):
    response = requests.get(API_URL, timeout=10)
    response.raise_for_status()

    for server in response.json():
        if server.get("ownerId") == owner_id:
            return server

    return None


# ---------- discord ----------

class Bot(discord.Client):
    def __init__(self):
        intents = discord.Intents.default()
        super().__init__(intents=intents)
        self.tree = app_commands.CommandTree(self)

    async def setup_hook(self):
        if GUILD_ID:
            guild = discord.Object(id=GUILD_ID)
            await self.tree.sync(guild=guild)
            print("Guild slash commands synced")
        else:
            await self.tree.sync()
            print("Global slash commands synced")


bot = Bot()


@bot.event
async def on_ready():
    print(f"Logged in as {bot.user} ({bot.user.id})")


# ---------- Slash Command ----------

@bot.tree.command(
    name="serverinfo",
    description="See information about a private server and be Able to join it.",
    guild=discord.Object(id=GUILD_ID) if GUILD_ID else None
)
@app_commands.describe(user_id="Roblox user ID")
async def server(interaction: discord.Interaction, user_id: int):
    await interaction.response.defer(thinking=True)

    server_data = find_server_by_owner_id(user_id)

    if not server_data:
        await interaction.followup.send(
            "No server found for this user ID."
        )
        return

    timestamp = int(time.time())

    embed = discord.Embed(
        title=server_data.get("serverName", "Unknown Server"),
        color=discord.Color.from_rgb(255, 255, 255)
    )

    private_server_id = server_data.get("privateServerId")

    embed.add_field(
        name="Private Server ID",
        value=f"`{private_server_id}`",
        inline=False
    )

    roblox_user_id = server_data.get("ownerId")
    roblox_username = server_data.get("ownerName", "Unknown")
    profile_url = f"https://www.roblox.com/users/{roblox_user_id}/profile"

    embed.add_field(
        name="Roblox User",
        value=f"[{roblox_username} ({roblox_user_id})]({profile_url})",
        inline=False
    )

    embed.add_field(
        name="Players",
        value=f"{server_data.get('currentPlayers')}/{server_data.get('maxPlayers')}",
        inline=True
    )

    embed.add_field(
        name="Last Update",
        value=f"<t:{timestamp}:R>\n<t:{timestamp}:F>",
        inline=True
    )

    avatar_url = server_data.get("ownerProfileUrl")
    if avatar_url:
        embed.set_thumbnail(url=avatar_url)

    # ---------- Join Service ----------

    join_url = (
        "https://www.roblox.com/games/start"
        f"?launchData=reservedServerId%3D{private_server_id}"
        "&placeId=7711635737"
    )

    # ---------- Join button ----------
    view = View()
    view.add_item(
        Button(
            label="Join the Server",
            style=discord.ButtonStyle.link,
            url=join_url
        )
    )

    await interaction.followup.send(embed=embed, view=view)


bot.run(TOKEN)

File: bot.js

///  Discord Bot for Emergency Hamburg Private Server Information
///  Coded by Unl1veabl3

const { Client, GatewayIntentBits, MessageFlags, ComponentType, ButtonStyle } = require('discord.js');
const axios = require('axios');

const client = new Client({
    intents: [GatewayIntentBits.Guilds]
});

const TOKEN = 'YOUR_BOT_TOKEN_HERE'; 
const API_URL = "https://api.emergency-hamburg.com/public/servers";

/// Place ID Of Emergency Hamburg, Needed for Join URL ///

const PLACE_ID = 7711635737;


async function findServerByOwnerId(userId) {
    try {
        const response = await axios.get(API_URL);
        const servers = response.data;
        
        if (!Array.isArray(servers)) {
            console.error('API returned non-array data:', servers);
            return null;
        }
        
        const server = servers.find(s => s.ownerId === userId);
        console.log(`Searching for userId ${userId}, found:`, server ? 'Yes' : 'No');
        
        return server;
    } catch (error) {
        console.error('Error fetching server data:', error.message);
        return null;
    }
}

client.once('ready', async () => {
    console.log(`Bot is online as ${client.user.tag}`);
    
    /// Here You can Change the Name of the Command and Description ///
    const commands = [
        {
            name: 'serverinfo',
            description: 'Show information about a private server',
            options: [
                {
                    name: 'user_id',
                    description: 'User ID',
                    type: 4,
                    required: true
                }
            ]
        }
    ];

    try {
        await client.application.commands.set(commands);
        console.log('Slash command registered');
    } catch (error) {
        console.error('Error registering commands:', error);
    }
});

client.on('interactionCreate', async interaction => {
    if (!interaction.isChatInputCommand()) return;

    if (interaction.commandName === 'serverinfo') {
        try {
            await interaction.deferReply();

            const userId = interaction.options.getInteger('user_id');
            const server = await findServerByOwnerId(userId);

            if (!server) {
                await interaction.editReply({
                    content: `No private server found for User ID: **${userId}**`
                });
                return;
            }

            const timestamp = Math.floor(Date.now() / 1000);
            const privateServerId = server.privateServerId || 'Unknown';
            const ownerId = server.ownerId;
            const ownerName = server.ownerName || 'Unknown User';
            const ownerProfileUrl = server.ownerProfileUrl;
            const joinUrl = `https://www.roblox.com/games/start?launchData=reservedServerId%3D${privateServerId}&placeId=${PLACE_ID}`;

            const components = [
                {
                    type: ComponentType.TextDisplay,
                    id: 1,
                    content: `# ${server.serverName || 'Unknown Server'}`
                },
                {
                    type: ComponentType.Separator,
                    id: 2
                }
            ];

            if (ownerProfileUrl) {
                components.push(
                    {
                        type: ComponentType.MediaGallery,
                        id: 3,
                        items: [
                            {
                                media: {
                                    url: ownerProfileUrl
                                },
                                description: `${ownerName}'s Profile`
                            }
                        ]
                    },
                    {
                        type: ComponentType.Separator,
                        id: 4
                    }
                );
            }

            const infoStartId = ownerProfileUrl ? 5 : 3;
            components.push(
                {
                    type: ComponentType.TextDisplay,
                    id: infoStartId,
                    content: `**Private Server ID**\n\`${privateServerId}\``
                },
                {
                    type: ComponentType.Separator,
                    id: infoStartId + 1
                },
                {
                    type: ComponentType.TextDisplay,
                    id: infoStartId + 2,
                    content: `**Server Owner**\n${ownerName}\n\`${ownerId}\``
                },
                {
                    type: ComponentType.Separator,
                    id: infoStartId + 3
                },
                {
                    type: ComponentType.TextDisplay,
                    id: infoStartId + 4,
                    content: `**Players Online**\n${server.currentPlayers || 0} / ${server.maxPlayers || 0}`
                },
                {
                    type: ComponentType.Separator,
                    id: infoStartId + 5
                },
                {
                    type: ComponentType.TextDisplay,
                    id: infoStartId + 6,
                    content: `**Last Update**\n<t:${timestamp}:R> • <t:${timestamp}:t>`
                },
                {
                    type: ComponentType.Separator,
                    id: infoStartId + 7
                },
                {
                    type: ComponentType.Section,
                    id: infoStartId + 8,
                    components: [
                        {
                            type: ComponentType.TextDisplay,
                            id: infoStartId + 9,
                            content: '​'
                        }
                    ],
                    accessory: {
                        type: ComponentType.Button,
                        style: ButtonStyle.Link,
                        label: 'Join Server',
                        url: joinUrl
                    }
                }
            );

            await interaction.editReply({
                flags: MessageFlags.IsComponentsV2,
                components: [
                    {
                        type: ComponentType.Container,
                        id: 100,
                        components: components,
                        spoiler: false
                    }
                ]
            });

        } catch (error) {
            console.error('Error in serverinfo command:', error);
            try {
                await interaction.editReply({
                    content: `An error occurred: ${error.message}`
                });
            } catch (e) {}
        }
    }
});

client.on('error', error => {
    console.error('Discord Client Error:', error);
});

process.on('unhandledRejection', error => {
    console.error('Unhandled Promise Rejection:', error);
});

client.login(TOKEN);